FoodFriends/app/src/main/java/com/example/foodfriends/activities/SignUpActivity.java

```diff
@@ -36,37 +39,67 @@ protected void onCreate(Bundle savedInstanceState) {
         etState = findViewById(R.id.etStateEntry);
         etCity = findViewById(R.id.etCityEntry);
         btnSignUp = findViewById(R.id.btnSignUp);
+        btnSignUp.setOnClickListener(this);
+    }
+
+
+    @Override
+    public void onClick(View v) {
+        switch(v.getId()){
 
-        btnSignUp.setOnClickListener(new View.OnClickListener() {
-            @Override
-            public void onClick(View v) {
+            case R.id.btnSignUp:
                 signUp();
-            }
-        });
+                break;
+        }
+
     }
 
+    /**
+     * Creates user object for the current user, with specified attributes
+     * **/
     private void signUp() {
-        ParseUser user = new ParseUser();
+        UserObservable user = new UserObservable();
         // Set the user's username and password, which can be obtained by a forms
         user.setUsername(etUsername.getText().toString());
         user.setPassword(etPassword.getText().toString());
-        user.put("state", etState.getText().toString());
-        user.put("city", etCity.getText().toString());
-        user.put("name", etName.getText().toString());
-
-        user.signUpInBackground(new SignUpCallback() {
-            @Override
-            public void done(ParseException e) {
-                if (e == null) {
-                    showAlert("Successful Sign Up!", "Welcome " + etName.getText().toString() +"!");
-                } else {
-                    ParseUser.logOut();
-                    Toast.makeText(SignUpActivity.this, e.getMessage(), Toast.LENGTH_LONG).show();
-                }
-            }
-        });
+        user.setState(etState.getText().toString());
+        user.setCity(etCity.getText().toString());
+        user.setName(etName.getText().toString());
+        switch (user.isValid()){
```

Comment:
Let's avoid switch case, instead have 

Map<Integer, String> userValidityStates = Map.ofEntries(entry(0,"Username must not be empty"), ...);